proofs/tree.go

```diff
@@ -342,45 +427,132 @@ func getPropertyNameFromProtobufTag(tag string) (name string, err error) {
-	value := f.messageValue.Field(index).Interface()
-	prop, err := getPropertyNameFromProtobufTag(tag)
-	if err != nil {
+func (f *messageFlattener) generateLeaves(propPrefix string, fcurrent *messageFlattener) (err error) {
+
+	if err = fcurrent.parseExtensions(); err != nil {
 		return err
 	}
 
-	// Check if the field has an exclude_from_tree option
-	if _, ok := f.excludedFields[prop]; ok {
-		return
+	for i := 0; i < fcurrent.messageValue.NumField(); i++ {
+		valueField := fcurrent.messageValue.Field(i)
+		valueType := reflect.TypeOf(valueField.Interface())
+		reflectValueFieldType := fcurrent.messageType.Field(i)
+
+		// Pointer dereference for correct type checks
+		if valueType.Kind() == reflect.Ptr {
+			valueType = valueType.Elem()
+		}
+
+		// Ignore fields starting with XXX_, those are protobuf internals
+		if strings.HasPrefix(reflectValueFieldType.Name, "XXX_") {
+			continue
+		}
+
+		tag := reflectValueFieldType.Tag.Get("protobuf")
+		prop, err1 := getPropertyNameFromProtobufTag(tag)
+		if err1 != nil {
+			return err1
+		}
+
+		// Check if the field has an exclude_from_tree option
+		if _, ok := fcurrent.excludedFields[prop]; ok {
+			continue
+		}
+
+		value := valueField.Interface()
+		reflectSaltsValue := fcurrent.saltsValue.FieldByName(reflectValueFieldType.Name)
+		salts := reflectSaltsValue.Interface()
+
+		value = DereferencePointer(value)
+		salts = DereferencePointer(salts)
+
+		if valueType.Kind() == reflect.Slice {
+			sliceValue := reflect.ValueOf(value)
+			if sliceValue.Type() == reflect.TypeOf([]uint8{}) { //Specific case where byte is internally represented as []uint8, but we want to treat it as a whole
+				salt := reflect.Indirect(fcurrent.saltsValue).FieldByName(reflectValueFieldType.Name).Interface().([]byte)
+				err = f.handleAppendLeaf(propPrefix, prop, value.([]byte), salt)
+			} else {
+				err = f.handleSlice(propPrefix, prop, fcurrent, i, sliceValue, salts)
+			}
+		} else if valueType.Kind() == reflect.Struct {
+			err = f.handleStruct(propPrefix, prop, valueField, valueType, reflectSaltsValue.Interface())
+		} else {
+			salt := reflect.Indirect(fcurrent.saltsValue).FieldByName(reflectValueFieldType.Name).Interface().([]byte)
+			err = f.handleAppendLeaf(propPrefix, prop, value, salt)
+		}
+
+		if err != nil {
+			return
+		}
+	}
+
+	return
+}
+
+func (f *messageFlattener) handleSlice(propPrefix string, prop string, fcurrent *messageFlattener, i int, sliceValue reflect.Value, salts interface{}) (err error) {
+	if !sliceValue.IsValid() {
+		return fmt.Errorf("Invalid value provided: %v", sliceValue)
+	}
+	ss := reflect.ValueOf(salts)
+
+	// Append length of slice as tree leaf
+	saltedFieldValue := fcurrent.saltsValue.FieldByName(fmt.Sprintf("%s%s", fcurrent.messageType.Field(i).Name, fcurrent.saltsLengthSuffix))
+	salt := saltedFieldValue.Interface().([]byte)
+	f.appendLeaf(fmt.Sprintf("%s%s.length", propPrefix, prop), strconv.Itoa(sliceValue.Len()), salt)
```

Comment:
I think it will force me to iterate over string list, to be able to concatenate, plus adding another variable. I rather keep it succint with `Sprintf`.